src/vfs2.c

```diff
@@ -1561,65 +1652,230 @@ int vfs2_read_wal(sqlite3_file *file,
+			txns[i].frames[j].commit = ByteGetBe32(fhdr.commit);
+			txns[i].frames[j].page = p;
+		}
+		if (from_wal_cur) {
+			vfs2_pseudo_read_end(file, read_lock);
+		}
+	}
 	return 0;
 }
 
-int vfs2_apply_uncommitted(sqlite3_file *file, const dqlite_vfs_frame *frames, unsigned len, struct vfs2_wal_slice *out)
+static int write_one_frame(struct entry *e, struct wal_frame_hdr hdr, void *data)
 {
-	(void)file;
-	(void)frames;
-	(void)len;
-	(void)out;
-	/*
-	struct vfs2_file *xfile = (struct vfs2_file *)file;
-	if (!(xfile->flags & SQLITE_OPEN_MAIN_DB)) {
-		return 1;
-	}
 	int rv;
 
-	unsigned *locks = xfile->db_shm.locks;
-	if (locks[VFS2_SHM_WRITE_LOCK] > 0) {
-		return 1;
+	sqlite3_int64 off = wal_offset_from_cursor(e->page_size, e->wal_cursor);
+	rv = e->wal_cur->pMethods->xWrite(e->wal_cur, &hdr, sizeof(hdr), off);
+	if (rv != SQLITE_OK) {
+		return rv;
 	}
-	locks[VFS2_SHM_WRITE_LOCK] = VFS2_EXCLUSIVE;
-
-	struct vfs2_salts salts;
-	rv = maybe_setup_or_swap_wal(xfile, len, &salts);
+	off += (sqlite3_int64)sizeof(hdr);
+	rv = e->wal_cur->pMethods->xWrite(e->wal_cur, data, (int)e->page_size, off);
 	if (rv != SQLITE_OK) {
-		return 1;
+		return rv;
+	}
+	e->wal_cursor += 1;
+	return SQLITE_OK;
+}
+
+static struct wal_hdr next_wal_hdr(struct entry *e)
+{
+	struct wal_hdr ret;
+	struct wal_hdr old = e->wal_cur_hdr;
+	BytePutBe32(native_magic(), ret.magic);
+	BytePutBe32(3007000, ret.version);
+	BytePutBe32(e->page_size, ret.page_size);
+	uint32_t ckpoint_seqno = ByteGetBe32(old.ckpoint_seqno);
+	BytePutBe32(ckpoint_seqno + 1, ret.ckpoint_seqno);
+	uint32_t salt1;
+       	if (ckpoint_seqno == 0) {
+		salt1 = get_salt1(old.salts) + 1;
+	} else {
+		e->common->orig->xRandomness(e->common->orig, sizeof(salt1), (void *)&salt1);
 	}
-	struct vfs2_file *wal = xfile->entry->wal;
+	BytePutBe32(salt1, ret.salts.salt1);
+	e->common->orig->xRandomness(e->common->orig, sizeof(ret.salts.salt2), (void *)&ret.salts.salt2);
+	return ret;
+}
 
-	// TODO assert that nothing is mapped => don't need to invalidate the WAL index
+static struct wal_frame_hdr txn_frame_hdr(struct entry *e, struct cksums sums, struct vfs2_wal_frame frame)
+{
+	struct wal_frame_hdr fhdr;
+
+	BytePutBe32(frame.page_number, fhdr.page_number);
+	BytePutBe32(frame.commit, fhdr.commit);
+	update_cksums(ByteGetBe32(e->wal_cur_hdr.magic), (const void *)(&fhdr), 8, &sums);
+	update_cksums(ByteGetBe32(e->wal_cur_hdr.magic), frame.page, e->page_size, &sums);
+	fhdr.salts = e->wal_cur_hdr.salts;
+	BytePutBe32(sums.cksum1, fhdr.cksum1);
+	BytePutBe32(sums.cksum2, fhdr.cksum2);
+	return fhdr;
+}
 
-	uint32_t page_size = atomic_load(&xfile->vfs_data->page_size);
-	assert(page_size > 0);
-	sqlite3_int64 ofst;
-	sqlite3_file *wal_cur = xfile->wal.wal_cur;
-	rv = wal_cur->xFileSize(wal_cur, &ofst);
-	if (rv != SQLITE_OK) {
-		return 1;
+int vfs2_apply_uncommitted(sqlite3_file *file, uint32_t page_size, const struct vfs2_wal_frame *frames, unsigned len, struct vfs2_wal_slice *out)
```

Comment:
Yes, it's technically redundant and I'm not absoultely committed to leaving it there. I included it so that this function can assert that the page size received in the AppendEntries message and the local page size are the same for this database.